pkg/fqdn/dnsproxy/proxy.go

```diff
@@ -151,22 +150,90 @@ type DNSProxy struct {
 	rejectReply int32
 }
 
+type matchNameMatcher struct {
+	mapping map[string]struct{}
+	key     *string
+}
+
+// regexEntry is a lookup entry used to cache a compiled regex
+// and how many policies it is being used by
+type regexEntry struct {
+	matcher        *regexp.Regexp
+	referenceCount int
+}
+
+// matchNamesEntry is a lookup entry used to cache a fqdn mapping
+// and how many policies it is being used by
+type matchNamesEntry struct {
+	matcher        *matchNameMatcher
+	referenceCount int
+}
+
 // perEPAllow maps EndpointIDs to ports + selectors + rules
-type perEPAllow map[uint64]portToSelectorAllow
+// It also contains indexes of all the regexps and the unique
+// fqdn maps in used, as well as how many rules refer to them.
+type perEPAllow struct {
+	allowMap map[uint64]portToSelectorAllow
+
+	// mapping from regexp pattern to regexp and reference count
+	patternMatchersByPattern map[string]*regexEntry
+
+	// mapping from list of matchNames concatenated with a "|" separator to
+	// map where matchNames can be looked up.
+	nameMatcherBySignature map[string]*matchNamesEntry
+}
+
+func newPerEPAllow() perEPAllow {
+	return perEPAllow{
+		allowMap:                 make(map[uint64]portToSelectorAllow),
+		patternMatchersByPattern: make(map[string]*regexEntry),
+		nameMatcherBySignature:   make(map[string]*matchNamesEntry),
+	}
+}
 
 // portToSelectorAllow maps port numbers to selectors + rules
-type portToSelectorAllow map[uint16]CachedSelectorREEntry
+type portToSelectorAllow map[uint16]CachedSelectorEntry
 
-// CachedSelectorREEntry maps port numbers to selectors to rules, mirroring
-// policy.L7DataMap but the DNS rules are compiled into a single regexp
-type CachedSelectorREEntry map[policy.CachedSelector]*regexp.Regexp
+// fqdnMatcher contain both a regexp used to match matcher to policy matchPatterns and
+// a map of matchNames to match against those.
+type fqdnMatcher struct {
+	patternMatcher *regexp.Regexp
+	nameMatcher    *matchNameMatcher
+}
+
+// CachedSelectorEntry maps port numbers to selectors to rules, mirroring
+// policy.L7DataMap but the DNS rules are compiled into a map of the matchNames and a regex
+// for the matchPatterns
+type CachedSelectorEntry map[policy.CachedSelector]fqdnMatcher
 
 // structure for restored rules that can be used while Cilium agent is restoring endpoints
 type perEPRestored map[uint64]restore.DNSRules
 
 // map from EP IPs to *Endpoint
 type restoredEPs map[string]*endpoint.Endpoint
 
+// matchFqdn checks if the fqdn is present in the internal map, and returns true
+// if it is.
+func (f *matchNameMatcher) matchFqdn(fqdn string) bool {
```

Comment:
```suggestion
func (f *matchNameMatcher) matchFQDN(fqdn string) bool {
```

Nit: We have precedence referring to "Fqdn" with full caps in this file and throughout the code base; let's make that consistent in this PR too.

Reply:
Nice! Thanks!